File: common/code/FrameStream/SlidingWindowFrameAssembler.cpp
// The MIT License (MIT)
// 
// Copyright (c) 2013-2014 Sound Metrics Corporation. All Rights Reserved.
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "SlidingWindowFrameAssembler.h"
#include "frame_stream.h"
#include <boost/asio/buffer.hpp>
#include <functional>

namespace {
// Execute the lambda expression at the end of the scope.
class scoped_guard : boost::noncopyable
{
    std::function<void()> lambda;

public:
    scoped_guard(std::function<void()> action)
        : lambda(action)
    {
    }

    ~scoped_guard()
    {
        try {
            lambda();
        }
        catch (...) {
            // Ignore.
        }
    }
};
}

namespace Aris {
namespace Network {

using namespace boost;
using namespace boost::asio;

namespace {
SlidingWindowFrameAssembler::Metrics
operator+(const SlidingWindowFrameAssembler::Metrics &a,
          const SlidingWindowFrameAssembler::Metrics &b) {
  auto m = a;
  m.uniqueFrameIndexCount += b.uniqueFrameIndexCount;
  m.finishedFrameCount += b.finishedFrameCount;
  m.completeFrameCount += b.completeFrameCount;
  m.skippedFrameCount += b.skippedFrameCount;
  m.totalExpectedFrameSize += b.totalExpectedFrameSize;
  m.totalReceivedFrameSize += b.totalReceivedFrameSize;
  m.totalPacketsReceived += b.totalPacketsReceived;
  m.totalPacketsAccepted += b.totalPacketsAccepted;
  m.totalPacketsIgnored += b.totalPacketsIgnored;
  m.invalidPacketCount += b.invalidPacketCount;

  return m;
}
}

SlidingWindowFrameAssembler::SlidingWindowFrameAssembler(
    std::function<void(int, int)> sendAck,
    std::function<void(FrameBuilder &)> onFrameFinished)
    : sendAck(sendAck), onFrameFinished(onFrameFinished), currentFrameIndex(-1),
      lastFinishedFrameIndex(-1), expectedDataOffset(0) {
  Metrics emptyMetrics = {};
  metrics = emptyMetrics;
}

void SlidingWindowFrameAssembler::ProcessPacket(const_buffer data) {

  bool acceptedPacket = false, invalidPacket = false;
  uint32_t skippedFrameCount = 0;
  scoped_guard updateMetrics([&]() {
    Metrics update = {};
    update.skippedFrameCount = skippedFrameCount;
    update.totalPacketsReceived = 1;
    update.totalPacketsAccepted = acceptedPacket ? 1 : 0;
    update.totalPacketsIgnored = acceptedPacket ? 0 : 1;
    update.invalidPacketCount = invalidPacket ? 1 : 0;
    UpdateMetrics(update);
  });

  recursive_mutex::scoped_lock lock(stateGuard);

  frame_stream::FramePart framePart;
  if (!framePart.ParseFromArray(buffer_cast<const uint8_t *>(data),
                                buffer_size(data))) {
    invalidPacket = true;
    // TODO log
    return;
  }

  const int incomingFrameIndex = framePart.frame_index();
  const int incomingDataOffset = framePart.data_offset();

  if (incomingFrameIndex > currentFrameIndex) {
    // sender moved on to the next frame
    Flush();
    skippedFrameCount = incomingFrameIndex - currentFrameIndex - 1;
    currentFrameIndex = incomingFrameIndex;
    expectedDataOffset = 0;
  } else if (incomingFrameIndex <= lastFinishedFrameIndex) {
    return; // duplicate packet from finished frame
  }

  if (!currentFrame) {
    if (incomingDataOffset == 0) {
      currentFrame = std::unique_ptr<FrameBuilder>(new FrameBuilder(
          incomingFrameIndex,
          buffer(framePart.header().c_str(), framePart.header().size()),
          buffer(framePart.data().c_str(), framePart.data().size()),
          framePart.total_data_size()));
      expectedDataOffset = framePart.data().size();
      acceptedPacket = true;
    } else {
      // Ack will go out asking for the first part of the frame to be resent.
    }
  } else {
    if (incomingDataOffset == expectedDataOffset) {
      currentFrame->AppendFrameData(
          incomingDataOffset,
          buffer(framePart.data().c_str(), framePart.data().size()));
      expectedDataOffset += framePart.data().size();
      acceptedPacket = true;
    } else {
      // Missed a part.
    }
  }

  // NOTE: we're always acking each packet for now; this should change when we
  // develop strategies for retrying packets.
  sendAck(incomingFrameIndex, expectedDataOffset);

  if (expectedDataOffset == framePart.total_data_size())
    Flush();
}

void SlidingWindowFrameAssembler::Flush() {
  recursive_mutex::scoped_lock lock(stateGuard);

  if (currentFrame) {
    std::unique_ptr<FrameBuilder> frame;
    frame.swap(currentFrame);
    assert(!currentFrame);

    UpdateMetrics(GetMetricsForFinishedFrame(*frame));

    lastFinishedFrameIndex = frame->FrameIndex();
    onFrameFinished(*frame);
  }
}

SlidingWindowFrameAssembler::Metrics SlidingWindowFrameAssembler::GetMetrics() {
  recursive_mutex::scoped_lock lock(metricsGuard);
  return metrics;
}

void SlidingWindowFrameAssembler::UpdateMetrics(const Metrics &update) {
  recursive_mutex::scoped_lock lock(metricsGuard);
  metrics = metrics + update;
}

/* static */
SlidingWindowFrameAssembler::Metrics
SlidingWindowFrameAssembler::GetMetricsForFinishedFrame(
    const FrameBuilder &frame) {
  Metrics metrics = {};
  metrics.uniqueFrameIndexCount = 1;
  metrics.finishedFrameCount = 1;
  metrics.completeFrameCount = frame.IsComplete() ? 1 : 0;
  metrics.totalExpectedFrameSize = frame.ExpectedSize();
  metrics.totalReceivedFrameSize = frame.BytesReceived();

  return metrics;
}
}
}

File: common/platform/SoundMetrics.Aris.Comms/SlidingWindowFrameAssembler.fs
// Copyright 2014-2018 Sound Metrics Corp. All Rights Reserved.

namespace SoundMetrics.Aris.Comms

open FrameStream
open Serilog
open System
open System.Threading.Tasks
open System.Threading.Tasks.Dataflow

[<AutoOpen>]
module private SlidingWindowFrameAssemblerLogging =

    let logVerboseFrameAssembly =
#if LOG_VERBOSE_FRAME_ASSEMBLY
        true
#else
        false
#endif

    let logSkippedFrame (currentFrame : FrameIndex) (incomingFrame : FrameIndex) (incomingDataOffset : uint32)
                        (gap : uint64) (cause : string) =
        if logVerboseFrameAssembly then
            Log.Verbose("Skipped frame currentFrame={CurrentFrame}; incomingFrame={IncomingFrame}; "
                + "incomingDataOffset={IncomingDataOffset}; gap={Gap}; cause={Cause}",
                currentFrame, incomingFrame, incomingDataOffset, gap, cause)

    let logDuplicatePacket (lastFinishedFrame : FrameIndex) (incomingFrame : FrameIndex) (incomingDataOffset : uint32) =
        if logVerboseFrameAssembly then
            Log.Verbose("Duplicate packet lastFinishedFrame={LastFinishedFrame}; incomingFrame={IncomingFrame}; incomingDataOffset={IncomingDataOffset}",
                lastFinishedFrame, incomingFrame, incomingDataOffset)

    let logMissedPacket (currentFrame : FrameIndex) (incomingFrame : FrameIndex) (expectedOffset : uint32) (incomingOffset : uint32) =
        if logVerboseFrameAssembly then
            Log.Verbose("Missed packet currentFrame={CurrentFrame}; incomingFrame={IncomingFrame}; "
                + "expectedOffset={ExpectedOffset}; incomingOffset={IncomingOffset}",
                currentFrame, incomingFrame, expectedOffset, incomingOffset)

    let logFinishedFrame (currentFrame : FrameIndex) (isComplete : bool) =
        if logVerboseFrameAssembly then
            Log.Verbose("Finished frame currentFrame={CurrentFrame}; isComplete={IsComplete}",
                currentFrame, isComplete)

    let logAcceptedPacket (currentFrame : FrameIndex) (incomingOffset : uint32) =
        if logVerboseFrameAssembly then
            Log.Verbose("Accepted packet currentFrame={CurrentFrame}; incomingOffset={IncomingOffset}",
                currentFrame, incomingOffset)

    let logResetFrameIndexes (lastFinishedFrame : FrameIndex) =
        Log.Information("Reset last finished frame index to {LastFinishedFrame}", lastFinishedFrame)

    let logReceivedWorkUnit (workUnitType : string) =
        if logVerboseFrameAssembly then
            Log.Verbose("Received work unit; type={WorkUnitType}", workUnitType)

    let logReceivingPacket (byteCount : int) =
        if logVerboseFrameAssembly then
            Log.Verbose("Receiving packet of {ByteCount} bytes", byteCount)

    let logCouldNotParseOrProcessFramePart (msg : string) stackTrace =
        Log.Warning("Couldn't parse or process FramePart: {Message}; {stackTrace}", msg, stackTrace)

    let logIncomingFrameIndexAdvanced (currentFI : FrameIndex) (incomingFI : FrameIndex) =
        if logVerboseFrameAssembly then
            Log.Verbose("Incoming frame index advanced from {CurrentFI} to {IncomingFI}",
                currentFI, incomingFI)

    let logFirstKnownFrame (fi : FrameIndex) =
        if logVerboseFrameAssembly then
            Log.Verbose("First known frame; fi={FrameIndex}", fi)

    let logSenderMovedOnToNextFrame (fi : FrameIndex) =
        if logVerboseFrameAssembly then
            Log.Verbose("Sender moved on to next frame; fi= {FrameIndex}", fi)

    let logResetCurrentFrameIndexTo (fi : FrameIndex) =
        if logVerboseFrameAssembly then
            Log.Verbose("Reset current frame index; fi={FrameIndex}", fi)

    let logReceivedPacketFromPreviousFrame (fi : FrameIndex) =
        if logVerboseFrameAssembly then
            Log.Verbose("Received packet from previous frame; fi={FrameIndex}", fi)

    let logSendingFramePartAck (fi : FrameIndex) (offset : uint32) =
        if logVerboseFrameAssembly then
            Log.Verbose("Sending frame part ack; fi={FrameIndex}; offset={Offset}", fi, offset)

    let logFrameComplete (fi : FrameIndex) (offset : uint32) =
        if logVerboseFrameAssembly then
            Log.Verbose("Frame complete; fi={FrameIndex}; offset={Offset}", fi, offset)

(*
    The sliding window frame assembler implements a simple sliding window method for use
    while assembling frames of packets received.
*)

type internal SendAck = FrameIndex -> uint32 -> unit
type internal FrameFinishedHandler = FrameAccumulator -> unit

module private SlidingWindowFrameAssemblerDetails =

    type WorkUnit =
    | Packet of data : byte array * timestamp : DateTimeOffset
    | Drain of TaskCompletionSource<unit> option

    /// Construct the TPL data graph that makes up the packet queue, which
    /// buffers then invokes processPacket. Returns the graph entry point
    /// and the disposable to tear down the graph.
    let makeWorkQueue (processWorkUnit: WorkUnit -> unit) =
        let pktQueue = BufferBlock<WorkUnit>()
        let pktProcessor = ActionBlock<WorkUnit>(processWorkUnit)
        let pktQueueLink = pktQueue.LinkTo(pktProcessor)
        // Return the entry into the graph and the disposable.
        pktQueue, pktQueueLink

    let onFrameIndexAdvanced currentFrameIndex incomingFrameIndex incomingDataOffset expectedDataOffset setSkipCount
                             flush =

        logIncomingFrameIndexAdvanced !currentFrameIndex incomingFrameIndex

        if !currentFrameIndex < 0 then
            // This is the first frame we're starting
            logFirstKnownFrame incomingFrameIndex
        else
            // sender moved on to the next frame
            logSenderMovedOnToNextFrame incomingFrameIndex

            let gap, _ = setSkipCount()
            logSkippedFrame !currentFrameIndex incomingFrameIndex incomingDataOffset gap
                            "sender moved on to the next frame"
            flush false

        System.Diagnostics.Trace.TraceInformation(sprintf "Reset currentFrameIndex to %d" !currentFrameIndex)
        logResetCurrentFrameIndexTo incomingFrameIndex
        currentFrameIndex := incomingFrameIndex
        expectedDataOffset := 0u

    let onReceivedPacketFromPreviousFrame incomingFrameIndex
                                          currentFrameIndex
                                          incomingDataOffset
                                          setSkipCount =

        // duplicate packet from finished frame
        logReceivedPacketFromPreviousFrame incomingFrameIndex

        let gap, _ = setSkipCount()
        logSkippedFrame !currentFrameIndex incomingFrameIndex incomingDataOffset gap
                        "duplicate packet from finished frame"

    let onRetrogradeFrameIndex incomingFrameIndex currentFrameIndex lastFinishedFrameIndex incomingDataOffset
                               flush =

        // Sonar may have reset; resync on first frame of a packet; we just checked
        // for incomingDataOffset <> 0 a few lines above.
        assert (incomingDataOffset = 0u)

        // Flush first (side effect), then reset indexes.
        flush true
        let newIndex = incomingFrameIndex
        currentFrameIndex := newIndex
        logResetFrameIndexes !lastFinishedFrameIndex

    let onFrameInProgress incomingFrameIndex currentFrameIndex incomingDataOffset expectedDataOffset acceptedPacket
                          (framePart: FramePart)
                          (accum: FrameAccumulator) =

        if incomingDataOffset = !expectedDataOffset then
            accum.AppendFrameData incomingDataOffset (framePart.Data.ToByteArray())
            expectedDataOffset := !expectedDataOffset + uint32 framePart.Data.Length
            acceptedPacket := true
            logAcceptedPacket incomingFrameIndex incomingDataOffset
        else
            // Missed a part. Could be caused by a missing packet or by ArisApp resetting and
            // starting over.
            logMissedPacket !currentFrameIndex incomingFrameIndex !expectedDataOffset incomingDataOffset

    let onNoFrameInProgress timestamp incomingFrameIndex currentFrameIndex incomingDataOffset expectedDataOffset acceptedPacket
                            currentFrame
                            (framePart: FramePart) =

        if incomingDataOffset = 0u then
            let accum = FrameAccumulator(
                                    incomingFrameIndex, framePart.Header.ToByteArray(), timestamp,
                                    framePart.Data.ToByteArray(), uint32 framePart.TotalDataSize)
            currentFrame := Some accum
            expectedDataOffset := uint32 framePart.Data.Length
            acceptedPacket := true
            logAcceptedPacket incomingFrameIndex incomingDataOffset
        else
            // Ack will go out asking for the first part of the frame to be resent.
            logMissedPacket !currentFrameIndex incomingFrameIndex 0u incomingDataOffset

open SlidingWindowFrameAssemblerDetails

/// Assembles packets into frames. See FrameStreamListener for policy regarding
/// packet retries and dropping partial frames.
[<Sealed(true)>]
type internal SlidingWindowFrameAssembler (sendAck: SendAck,
                                           onFrameFinished: FrameFinishedHandler) =
    let disposed = ref false
    let currentFrameIndex = ref -1
    let lastFinishedFrameIndex = ref -1
    let expectedDataOffset = ref 0u
    let currentFrame: FrameAccumulator option ref = ref None
    let metrics = ref ProtocolMetrics.Empty

    let stateGuard = Object()
    let metricsGuard = Object()

    let updateMetrics update =
        lock metricsGuard (fun () -> metrics := !metrics + update)

    let getMetricsForFinishedFrame (frame: FrameAccumulator) =
        { ProtocolMetrics.Empty with
            UniqueFrameIndexCount = 1UL;
            ProcessedFrameCount = 1UL;
            CompleteFrameCount = if frame.IsComplete then 1UL else 0UL;
            TotalExpectedFrameSize = uint64 frame.ExpectedSize;
            TotalReceivedFrameSize = uint64 frame.BytesReceived; }

    /// Assigns a new frame index to the frame (mutation!)
    let assignNewFrameIndex = (* FrameAccumulator -> unit *)
        // Close over state monotonicFrameIndex.
        let monotonicFrameIndex = ref 0 // Frame numbers start at zero. Often this is overwritten
                                        // (e.g. during recording) but we start at zero.
        fun (frame: FrameAccumulator) ->
            let idx = !monotonicFrameIndex
            frame.SetFrameIndex idx
            monotonicFrameIndex := idx + 1

    let flush clearFrameIndex =
        lock  stateGuard (fun () -> 
            match !currentFrame with
            | Some frame ->
                currentFrame := None
                assignNewFrameIndex frame
                updateMetrics (getMetricsForFinishedFrame frame)
                lastFinishedFrameIndex := frame.FrameIndex
                logFinishedFrame frame.FrameIndex frame.IsComplete
                currentFrameIndex := !currentFrameIndex + 1
                onFrameFinished frame
            | None -> ()

            if clearFrameIndex then
                lastFinishedFrameIndex := -1
        )

    // Process a packet received on the frame stream.
    let receivePacket (data: byte array) (timestamp : DateTimeOffset) =
        logReceivingPacket data.Length

        // Maintain partial metrics for updating metrics below.
        let acceptedPacket = ref false
        let skippedFrameCount = ref 0UL
        let parsedOkay = ref false

        try // finally
            try // with
                let framePart = FramePart.Parser.ParseFrom(data)
                parsedOkay := true

                lock stateGuard (fun () ->
                    let incomingFrameIndex = framePart.FrameIndex
                    let incomingDataOffset = uint32 framePart.DataOffset

                    let setSkipCount () =
                        let gap = uint64(incomingFrameIndex - !currentFrameIndex - 1)
                        let newValue = !skippedFrameCount + gap
                        skippedFrameCount := newValue
                        gap, newValue

                    // The sonar is responsible for deciding  when to move on to the next frame
                    // so regardless of our state we move on with it.
                    if incomingFrameIndex > !currentFrameIndex then
                        onFrameIndexAdvanced currentFrameIndex incomingFrameIndex incomingDataOffset expectedDataOffset
                                             setSkipCount flush

                    if incomingFrameIndex < !currentFrameIndex && incomingDataOffset <> 0u then
                        onReceivedPacketFromPreviousFrame incomingFrameIndex currentFrameIndex incomingDataOffset
                                                          setSkipCount
                    else
                        if incomingFrameIndex < !currentFrameIndex then
                            onRetrogradeFrameIndex incomingFrameIndex currentFrameIndex lastFinishedFrameIndex incomingDataOffset
                                                   flush

                        match !currentFrame with
                        | Some cf ->
                            onFrameInProgress incomingFrameIndex currentFrameIndex incomingDataOffset expectedDataOffset
                                              acceptedPacket framePart cf

                        | None ->
                            onNoFrameInProgress timestamp incomingFrameIndex currentFrameIndex incomingDataOffset expectedDataOffset
                                                acceptedPacket currentFrame framePart

                        // NOTE: we're always acking each packet for now; this should change when we
                        // develop strategies for retrying packets.
                        logSendingFramePartAck incomingFrameIndex !expectedDataOffset
                        sendAck incomingFrameIndex !expectedDataOffset

                        if !expectedDataOffset = uint32 framePart.TotalDataSize then
                            logFrameComplete incomingFrameIndex !expectedDataOffset
                            flush false
                    )
            with
                e -> logCouldNotParseOrProcessFramePart e.Message e.StackTrace
        finally
            updateMetrics { ProtocolMetrics.Empty with
                                SkippedFrameCount = skippedFrameCount.Value
                                TotalPacketsReceived = 1UL
                                TotalPacketsAccepted = if acceptedPacket.Value then 1UL else 0UL
                                UnparsablePackets = if parsedOkay.Value then 0UL else 1UL }


    let processWorkUnit workUnit =
        logReceivedWorkUnit (workUnit.GetType().Name)

        match workUnit with
        | Packet (packetData, timestamp) -> receivePacket packetData timestamp
        | Drain tcs ->  match tcs with
                        | Some t -> t.SetResult(())
                        | None -> ()
        
    let workQueue, pktQueueLink = makeWorkQueue processWorkUnit

    interface IDisposable with
        member __.Dispose() =
            Dispose.theseWith disposed
                [ pktQueueLink ]
                (fun () ->
                    workQueue.Complete()
                    workQueue.Completion.Wait())
    member s.Dispose() = (s :> IDisposable).Dispose()

    /// Flushes the current frame and resets to expect any frame index next.
    member __.Flush () = flush true
    member __.Metrics = !metrics
    member __.ProcessPacket packetData = workQueue.Post(Packet packetData)
    member __.DrainAsync () =
        let tcs = TaskCompletionSource<unit>()
        workQueue.Post(Drain (Some tcs)) |> ignore
        tcs.Task
